Automattic/woocommerce-payments
# New WCPay Subscription migrator class and a function to migrate a single subscription

Fixes #6524 

#### Updates since opening PR

- Introduced a feature flag `_wcpay_feature_allow_subscription_migrations`
- Refactored the migration process to use a try/catch approach
- If the subscription is already migrated, exit the migration rather than double-check it has been canceled at Stripe. The reason for this change is that it simplifies the migration code a lot for very little impact. If you need to migrate again, just rename the `_migrated_wcpay_subscription_id` meta key to `_wcpay_subscription_id` and go again.
- Moved the setting of the `_migrated_wcpay_subscription_id` meta to be the very final step of migration so we can be certain this meta is set after everything is finished.

---

#### Changes proposed in this Pull Request









This PR is the first of many PRs to implement migrating subscriptions from a WCPay Subscription (which uses the Stripe Billing engine) over to a tokenized subscription.

Included in this PR:
1. Introducing a new `WC_Payments_Subscriptions_Migrator` class which will house all of the code to handle all things WCPay subscription migration related
2. A new function `migrate_wcpay_subscription()` to handle migrating a single subscription

This PR doesn't o anything by itself and can be mostly treated as a WIP however, the idea is that this new `migrate_wcpay_subscription()` will be hooked onto a single scheduled action with a subscription ID and will handle migrating the subscription and logging all the things.

> **Note** - For debugging purposes I have implemented a dummy `log()` function which just calls `error_log()` for the time being. At a later stage this function will be fleshed out once we start tackling https://github.com/Automattic/woocommerce-payments/issues/6528

#### Migration process

I have split the migration process into the following steps:

1. Validate the subscription ID can be migrated
   1. Confirm the WC Subscription extension is active otherwise migrated subscriptions will be processed as a manual renewal.
   2. Check if the site is not in staging mode or a duplicate site
   3. Confirm the subscription ID exists in the DB and is a subscription.
   4. Check if the subscription has already been migrated before
3. Attempt to cancel the subscription in Stripe
   1. Only cancel the subscription at Stripe if it has an active status. I have considered active subscriptions as ones with statuses: `active`, `past_due`, `trialing`, `paused` ([Stripe API ref](https://stripe.com/docs/api/subscriptions/object#subscription_object-status))
4. Remove the `_wcpay_subscription_id` metadata and store it in `_migrated_wcpay_subscription_id` 
5. Add an order note to the subscription to record the migration as successful



#### Testing instructions





1. Set `_wcpay_feature_allow_subscription_migrations` in wp-options to `1`
2. Deactive the WC Subscriptions plugin and enable the WCPay Subscriptions feature
3. Purchase a couple of subscription products to create multiple WCPay Subscriptions
4. Enable the WC Subscriptions extension
5. With one of the subscription IDs run the following snippet to kick-off the migration:
```
$migrator = new WC_Payments_Subscriptions_Migrator( WC_Payments::get_payments_api_client() );
$migrator->migrate_wcpay_subscription( 1234 );
```
6. Open your PHP debug log and you should see some logs like:
```
Migrating subscription #123 (sub_XXXX)
---- Subscription at Stripe has "active" status. Canceling the subscription.
---- Subscription successfully canceled at Stripe.
---- SUCCESS: Subscription migrated.
```
7. Visit the stripe dashboard and confirm the subscription is canceled
8. Visit the Edit Subscription page and take note of the new order note
9. Process a renewal for this subscription
10. Confirm the subscription renewal payment is successful


> **Warning** - there's currently an issue where an existing subscription that is migrated may not have a pending scheduled action to process the payment. To fix this we'll need to make sure the tokenized subscription is going to renew properly after it's migrated.

-------------------


**Post merge**
